File: anyharness/crates/anyharness-contract/src/v1/hosting.rs
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

// ---------------------------------------------------------------------------
// Current pull request
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct CurrentPullRequestResponse {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pull_request: Option<PullRequestSummary>,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct PullRequestSummary {
    pub number: u64,
    pub title: String,
    pub url: String,
    pub state: PullRequestState,
    pub draft: bool,
    pub head_branch: String,
    pub base_branch: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum PullRequestState {
    Open,
    Closed,
    Merged,
}

// ---------------------------------------------------------------------------
// Create pull request
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct CreatePullRequestRequest {
    pub title: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub body: Option<String>,
    pub base_branch: String,
    #[serde(default)]
    pub draft: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct CreatePullRequestResponse {
    pub pull_request: PullRequestSummary,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub manual_url: Option<String>,
}

// ---------------------------------------------------------------------------
// Repo-root pull-request statuses
// ---------------------------------------------------------------------------

/// Reduced CI-check rollup state for a pull request's head commit.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum PullRequestChecksState {
    None,
    Pending,
    Passing,
    Failing,
}

/// Reduced review decision for a pull request.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum PullRequestReviewDecision {
    None,
    Approved,
    ChangesRequested,
}

/// [`PullRequestSummary`] plus reduced check-rollup and review-decision
/// fields. Both are optional so new clients deserialize old daemons (and
/// vice versa); absent maps to "none" at the domain layer.
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct BranchPullRequestSummary {
    pub number: u64,
    pub title: String,
    pub url: String,
    pub state: PullRequestState,
    pub draft: bool,
    pub head_branch: String,
    pub base_branch: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub checks: Option<PullRequestChecksState>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub review_decision: Option<PullRequestReviewDecision>,
}

/// One queried head branch. `pullRequest` null/absent means the branch WAS
/// queried and has no PR (authoritative none). Branches missing from
/// [`RepoPullRequestStatusesResponse::entries`] were not queried (unknown).
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct BranchPullRequestStatus {
    pub head_branch: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pull_request: Option<BranchPullRequestSummary>,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct RepoPullRequestStatusesResponse {
    pub entries: Vec<BranchPullRequestStatus>,
    pub fetched_at: String,
}

File: anyharness/sdk/src/types/hosting.ts
import type { components } from "../generated/openapi.js";

export type PullRequestState = components["schemas"]["PullRequestState"];
export type PullRequestSummary = components["schemas"]["PullRequestSummary"];
export type CurrentPullRequestResponse = components["schemas"]["CurrentPullRequestResponse"];
export type CreatePullRequestRequest = components["schemas"]["CreatePullRequestRequest"];
export type CreatePullRequestResponse = components["schemas"]["CreatePullRequestResponse"];

// ---------------------------------------------------------------------------
// Repo-root pull-request statuses
// ---------------------------------------------------------------------------
// Hand-written mirrors of the daemon v1 hosting contract additions backing
// GET /v1/repo-roots/{repo_root_id}/hosting/pull-requests. Once the daemon
// endpoint lands and `pnpm generate` regenerates openapi.ts, these become
// aliases of the generated schemas (shapes must stay identical).

export type PullRequestChecksState = "none" | "pending" | "passing" | "failing";

export type PullRequestReviewDecision = "none" | "approved" | "changes_requested";

/**
 * PullRequestSummary as served by daemons that already reduce check rollups
 * and review decisions. Both fields are optional so new clients deserialize
 * old daemons (and vice versa); absent maps to "none" at the domain layer.
 */
export type BranchPullRequestSummary = PullRequestSummary & {
  checks?: PullRequestChecksState | null;
  reviewDecision?: PullRequestReviewDecision | null;
};

/**
 * One queried head branch. `pullRequest` null/absent means the branch WAS
 * queried and has no PR (authoritative none). Branches missing from
 * `RepoPullRequestStatusesResponse.entries` were not queried (unknown).
 */
export interface BranchPullRequestStatus {
  headBranch: string;
  pullRequest?: BranchPullRequestSummary | null;
}

export interface RepoPullRequestStatusesResponse {
  entries: BranchPullRequestStatus[];
  fetchedAt: string;
}
